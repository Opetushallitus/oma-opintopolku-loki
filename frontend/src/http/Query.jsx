import React from 'react'
import PropTypes from 'prop-types'
import http from './http'

class Query extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      data: null,
      error: null,
      pending: true
    }
  }

  componentDidMount () {
    this.doQuery()
  }

  async doQuery () {
    const { url, method, body } = this.props

    try {
      const { data } = await http.request(url, method, body)
      this.setState({ data, pending: false })
    } catch (error) {
      console.error(error)
      this.setState({ error, pending: false })
    }
  }

  render () {
    return this.props.children(this.state)
  }
}

Query.defaultProps = {
  method: 'get',
  body: {}
}

Query.propTypes = {
  url: PropTypes.string.isRequired,
  method: PropTypes.string,
  body: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array
  ]),
  children: PropTypes.func.isRequired
}

export default Query
