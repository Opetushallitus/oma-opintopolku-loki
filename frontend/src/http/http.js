import axios from 'axios'

const client = axios.create({
  baseURL: process.env.API_BASE_URL
})

const get = async url => client.get(url)
const post = async (url, data) => client.post(url, data)

const request = async (url, method, data = {}) => {
  switch (method) {
    case 'GET':
    case 'get':
      return get(url)
    case 'POST':
    case 'post':
      return post(url, data)
    default:
      throw new Error(`Unsupported HTTP method: ${method}`)
  }
}

export default {
  get,
  post,
  request
}
