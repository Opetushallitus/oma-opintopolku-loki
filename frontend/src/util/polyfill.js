import 'core-js/features/promise'
import 'core-js/features/string/starts-with'
