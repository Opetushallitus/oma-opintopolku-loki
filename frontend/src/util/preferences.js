import Cookies from 'js-cookie'

export const lang = Cookies.get('lang') || 'fi'
