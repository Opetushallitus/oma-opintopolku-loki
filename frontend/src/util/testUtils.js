/* istanbul ignore file */
import renderer from 'react-test-renderer'

export function matchesSnapshot (component) {
  const renderedComponent = renderer.create(component)
  const componentJSON = renderedComponent.toJSON()
  expect(componentJSON).toMatchSnapshot()
  renderedComponent.unmount()
}
