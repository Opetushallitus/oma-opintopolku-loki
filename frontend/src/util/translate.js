import Cookies from 'js-cookie'

export default key => {
  const lang = Cookies.get('lang') || 'fi'
  const translations = window.translationsMap || {}
  const versions = translations[key] || {}
  const translated = versions[lang]

  if (!translated) {
    console.error(`Translation missing for language ${lang}: ${key}`)
    return key
  }

  return translated
}
