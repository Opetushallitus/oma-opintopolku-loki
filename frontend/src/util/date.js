import moment from 'moment'
import { lang } from 'util/preferences'

moment.locale(lang)

export const isoStringToDate = isoString => moment(isoString).format('l, LT')
