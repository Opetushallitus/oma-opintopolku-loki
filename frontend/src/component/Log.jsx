import React from 'react'
import { lensProp, map, over, view } from 'ramda'
import Query from 'http/Query'
import { lang } from 'util/preferences'
import { AlertText } from 'ui/typography'
import constants from 'ui/constants'
import t from 'util/translate'
import Organization from 'component/organization/Organization'
import styled from 'styled-components'

const organizationLens = lensProp('organizations')
const nameLens = lensProp('name')
const oidLens = lensProp('oid')

const NotificationText = styled.div`
  margin: 4.429rem 0;
  font-size: ${constants.font.size.xl};
  letter-spacing: 0.0214rem;
`

const getTranslatedName = organizationName => organizationName[lang] || ''

const Log = () => (
  <Query url='auditlog'>
    {({ data, error, pending }) => {
      if (error) return <AlertText>{t`Tapahtui odottamaton virhe, emmekä juuri nyt pysty näyttämään tietoja.`}</AlertText>
      if (pending) return <div>{t`Tietoja haetaan`}</div>
      if (!Array.isArray(data) || !data.length) return <NotificationText>{t`Sivu on tyhjä, koska tietojasi ei ole vielä käytetty`}</NotificationText>

      const translatedOrganizations = map(over(organizationLens, map(over(nameLens, getTranslatedName))))(data)

      return translatedOrganizations.map(({ organizations, timestamps }) => {
        const key = map(view(oidLens), organizations).join(',')

        return (
          <Organization
            key={key}
            organizationAlternatives={organizations}
            timestamps={timestamps}
          />
        )
      })
    }}
  </Query>
)

export default Log
