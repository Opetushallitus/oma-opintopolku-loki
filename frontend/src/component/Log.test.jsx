import React from 'react'
import Log from './Log'
import { create } from 'react-test-renderer'
import http from '../http/http'
import * as preferences from 'util/preferences'

describe('Log', () => {
  const logData = [
    {
      organizations: [
        {
          oid: '1.1.111.111.11.11111111111',
          name: { fi: 'Organisaatio 1', sv: 'Organisation 1', en: 'Organization 1' }
        }
      ],
      timestamps: ['2018-09-19T12:05:26.432+03', '2018-09-20T16:01:44.543+03']
    },
    {
      organizations: [
        {
          oid: '2.2.222.222.22.22222222222',
          name: { fi: 'Test 2', sv: 'Test 2', en: 'Test 2' }
        }
      ],
      timestamps: ['2018-07-19T21:38:35.104+03', '2018-08-26T18:56:17.437+03', '2017-06-07T23:58:55.136+03']
    }
  ]

  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should render with empty log', (done) => {
    http.request = jest.fn(() => Promise.resolve({ data: [] }))

    const renderer = create(<Log/>)

    setTimeout(() => {
      const componentJSON = renderer.toJSON()
      expect(componentJSON).toMatchSnapshot()
      renderer.unmount()

      expect(http.request).toHaveBeenCalledTimes(1)

      done()
    })
  })

  it('should render with non-empty log', (done) => {
    http.request = jest.fn(() => Promise.resolve({ data: logData }))

    const renderer = create(<Log/>)

    setTimeout(() => {
      const componentJSON = renderer.toJSON()
      expect(componentJSON).toMatchSnapshot()
      renderer.unmount()

      expect(http.request).toHaveBeenCalledTimes(1)

      done()
    })
  })

  it('should render with non-empty log and language en', (done) => {
    http.request = jest.fn(() => Promise.resolve({ data: logData }))
    const originalLang = preferences.lang
    preferences.lang = 'en'

    const renderer = create(<Log/>)

    setTimeout(() => {
      const componentJSON = renderer.toJSON()
      expect(componentJSON).toMatchSnapshot()
      renderer.unmount()
      preferences.lang = originalLang

      expect(http.request).toHaveBeenCalledTimes(1)

      done()
    })
  })

  it('should render with non-empty log and language sv', (done) => {
    http.request = jest.fn(() => Promise.resolve({ data: logData }))
    const originalLang = preferences.lang
    preferences.lang = 'sv'

    const renderer = create(<Log/>)

    setTimeout(() => {
      const componentJSON = renderer.toJSON()
      expect(componentJSON).toMatchSnapshot()
      renderer.unmount()
      preferences.lang = originalLang

      expect(http.request).toHaveBeenCalledTimes(1)

      done()
    })
  })

  it('should render with non-empty log and organization with no English name', (done) => {
    const data = [
      {
        organizations: [
          {
            oid: '1.1.111.111.11.11111111111',
            name: { fi: 'Organisaatio 1', sv: 'Organisation 1' }
          }
        ],
        timestamps: ['2018-09-19T12:05:26.432+03']
      }
    ]
    http.request = jest.fn(() => Promise.resolve({ data }))
    const originalLang = preferences.lang
    preferences.lang = 'en'

    const renderer = create(<Log/>)

    setTimeout(() => {
      const componentJSON = renderer.toJSON()
      expect(componentJSON).toMatchSnapshot()
      renderer.unmount()
      preferences.lang = originalLang

      expect(http.request).toHaveBeenCalledTimes(1)

      done()
    })
  })

  it('should render alert in case of error in query', (done) => {
    http.request = jest.fn(() => {
      const error = new Error('Request failed!')
      return Promise.reject(error)
    })
    console.error = jest.fn()

    const renderer = create(<Log/>)

    setTimeout(() => {
      const componentJSON = renderer.toJSON()
      expect(componentJSON).toMatchSnapshot()
      renderer.unmount()

      expect(http.request).toHaveBeenCalledTimes(1)
      expect(console.error).toHaveBeenCalledTimes(1)

      done()
    })
  })
})
