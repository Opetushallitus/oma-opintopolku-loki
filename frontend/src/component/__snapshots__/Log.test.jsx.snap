// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log should render alert in case of error in query 1`] = `
<strong
  className="typography__AlertText-sc-1yq3b4j-4 jglpck"
>
  Tapahtui odottamaton virhe, emmekä juuri nyt pysty näyttämään tietoja.
</strong>
`;

exports[`Log should render with empty log 1`] = `
<div>
  <div
    className="Log__NotificationText-sc-1asjtoe-0 jHbnhs"
  >
    <p>
      Tällä sivulla näytetään toimijat, jotka ovat käyttäneet opintotietojasi.
    </p>
    <p>
      Sivu on tyhjä, koska tietojasi ei ole vielä käytetty.**
    </p>
  </div>
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    *
     
    Koskee vain opintotietoja, ei hakemustietoja. Lisätietoa hakemusten henkilötietojen käsittelyn perusteista sekä henkilötietojen luovuttamisesta saat 
    <a
      className="typography__Link-sc-1yq3b4j-5 jcVYbI"
      href="https://opintopolku.fi/wp/fi/tietosuojaseloste/"
      target="_blank"
    >
      <test-file-stub
        className="ExternalLink__PositionedLinkIcon-sc-1in9m5g-0 skkip"
      />
      <b
        className="typography__Bold-sc-1yq3b4j-6 lbwfzX"
      >
        tietosuojaselosteesta
      </b>
    </a>
    .
  </div>
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    **
     
    Tietojen käyttäjät näytetään 14.11.2018 jälkeiseltä ajalta.
  </div>
</div>
`;

exports[`Log should render with non-empty log 1`] = `
Array [
  <div
    className="OrganizationsHeader__Container-sc-1dz6xpo-0 iCgOHG"
  >
    Tietojen käyttäjä**
  </div>,
  <div
    className="Expander__ExpanderContainer-fukmjx-0 bFzkEe"
  >
    <button
      aria-pressed={false}
      className="Expander__ExpanderTitle-fukmjx-2 hrNEBO"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
        className="Expander__ExpanderPrefix-fukmjx-1 fgEqwg"
      >
        +
      </span>
      Organisaatio 1
    </button>
  </div>,
  <div
    className="Expander__ExpanderContainer-fukmjx-0 bFzkEe"
  >
    <button
      aria-pressed={false}
      className="Expander__ExpanderTitle-fukmjx-2 hrNEBO"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
        className="Expander__ExpanderPrefix-fukmjx-1 fgEqwg"
      >
        +
      </span>
      Test 2
    </button>
  </div>,
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    *
     
    Koskee vain opintotietoja, ei hakemustietoja. Lisätietoa hakemusten henkilötietojen käsittelyn perusteista sekä henkilötietojen luovuttamisesta saat 
    <a
      className="typography__Link-sc-1yq3b4j-5 jcVYbI"
      href="https://opintopolku.fi/wp/fi/tietosuojaseloste/"
      target="_blank"
    >
      <test-file-stub
        className="ExternalLink__PositionedLinkIcon-sc-1in9m5g-0 skkip"
      />
      <b
        className="typography__Bold-sc-1yq3b4j-6 lbwfzX"
      >
        tietosuojaselosteesta
      </b>
    </a>
    .
  </div>,
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    **
     
    Tietojen käyttäjät näytetään 14.11.2018 jälkeiseltä ajalta.
  </div>,
]
`;

exports[`Log should render with non-empty log and language en 1`] = `
Array [
  <div
    className="OrganizationsHeader__Container-sc-1dz6xpo-0 iCgOHG"
  >
    User of data**
  </div>,
  <div
    className="Expander__ExpanderContainer-fukmjx-0 bFzkEe"
  >
    <button
      aria-pressed={false}
      className="Expander__ExpanderTitle-fukmjx-2 hrNEBO"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
        className="Expander__ExpanderPrefix-fukmjx-1 fgEqwg"
      >
        +
      </span>
      Organization 1
    </button>
  </div>,
  <div
    className="Expander__ExpanderContainer-fukmjx-0 bFzkEe"
  >
    <button
      aria-pressed={false}
      className="Expander__ExpanderTitle-fukmjx-2 hrNEBO"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
        className="Expander__ExpanderPrefix-fukmjx-1 fgEqwg"
      >
        +
      </span>
      Test 2
    </button>
  </div>,
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    *
     
    This applies only to study-related information. Please check the 
    <a
      className="typography__Link-sc-1yq3b4j-5 jcVYbI"
      href="https://studyinfo.fi/wp2/en/register/"
      target="_blank"
    >
      <test-file-stub
        className="ExternalLink__PositionedLinkIcon-sc-1in9m5g-0 skkip"
      />
      <b
        className="typography__Bold-sc-1yq3b4j-6 lbwfzX"
      >
        register description
      </b>
    </a>
     if you want to know how the information you have given in joint/separate applications is used.
  </div>,
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    **
     
    Showing organisations that have used your data after 14.11.2018.
  </div>,
]
`;

exports[`Log should render with non-empty log and language sv 1`] = `
Array [
  <div
    className="OrganizationsHeader__Container-sc-1dz6xpo-0 iCgOHG"
  >
    Uppgiftsanvändare**
  </div>,
  <div
    className="Expander__ExpanderContainer-fukmjx-0 bFzkEe"
  >
    <button
      aria-pressed={false}
      className="Expander__ExpanderTitle-fukmjx-2 hrNEBO"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
        className="Expander__ExpanderPrefix-fukmjx-1 fgEqwg"
      >
        +
      </span>
      Organisation 1
    </button>
  </div>,
  <div
    className="Expander__ExpanderContainer-fukmjx-0 bFzkEe"
  >
    <button
      aria-pressed={false}
      className="Expander__ExpanderTitle-fukmjx-2 hrNEBO"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
        className="Expander__ExpanderPrefix-fukmjx-1 fgEqwg"
      >
        +
      </span>
      Test 2
    </button>
  </div>,
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    *
     
    Gäller studieuppgifter, dvs. inte uppgifter som ingår i ansökningar. I 
    <a
      className="typography__Link-sc-1yq3b4j-5 jcVYbI"
      href="https://studieinfo.fi/wp/sv/dataskyddsbeskrivning/"
      target="_blank"
    >
      <test-file-stub
        className="ExternalLink__PositionedLinkIcon-sc-1in9m5g-0 skkip"
      />
      <b
        className="typography__Bold-sc-1yq3b4j-6 lbwfzX"
      >
        dataskyddsbeskrivningen
      </b>
    </a>
     får du information om grunderna för användning av personuppgifter i ansökningarna och hur de kan överlåtas.
  </div>,
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    **
     
    Här visas aktörerna som använt dina studieuppgifterna efter 14.11.2018.
  </div>,
]
`;

exports[`Log should render with non-empty log and organization with no English name 1`] = `
Array [
  <div
    className="OrganizationsHeader__Container-sc-1dz6xpo-0 iCgOHG"
  >
    User of data**
  </div>,
  <div
    className="Expander__ExpanderContainer-fukmjx-0 bFzkEe"
  >
    <button
      aria-pressed={false}
      className="Expander__ExpanderTitle-fukmjx-2 hrNEBO"
      onClick={[Function]}
    >
      <span
        aria-hidden="true"
        className="Expander__ExpanderPrefix-fukmjx-1 fgEqwg"
      >
        +
      </span>
      Organisaatio 1
    </button>
  </div>,
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    *
     
    This applies only to study-related information. Please check the 
    <a
      className="typography__Link-sc-1yq3b4j-5 jcVYbI"
      href="https://studyinfo.fi/wp2/en/register/"
      target="_blank"
    >
      <test-file-stub
        className="ExternalLink__PositionedLinkIcon-sc-1in9m5g-0 skkip"
      />
      <b
        className="typography__Bold-sc-1yq3b4j-6 lbwfzX"
      >
        register description
      </b>
    </a>
     if you want to know how the information you have given in joint/separate applications is used.
  </div>,
  <div
    className="Footnote__Container-x809r1-0 cYUoyK"
  >
    **
     
    Showing organisations that have used your data after 14.11.2018.
  </div>,
]
`;
